### ISSUE TITLE
Meta-issue for kostiantyn-kucher review of JOSS submission

### ISSUE THREAD

**Issue by kostiantyn-kucher** [MrShoenel/metrics-as-scores#15]
This is an issue created for my review of the JOSS manuscript, as discussed at https://github.com/openjournals/joss-reviews/issues/4913

I am glad to state that I can cross off all the items from the review checklist (please take my previous comments regarding the potential conflicts of interest into account, though), but I have some additional comments as part of my review that I would like to mention next.

In my opinion, the proposed tool indeed supports the functionality claimed in the documentation and the JOSS manuscript, while the underlying methodology and its validity (especially in the context of software quality metrics) are discussed in detail in the prior publication by the authors (Hönel et al., 2022). The statement of need, the general description of the tool, and information regarding installation, license, and contribution procedures are provided within the manuscript and the repository, respectively. I also tried downloading, building, running the tests, and using the tool locally with the Iris data set, with my experience corresponding to the expectations set by the documentation (the web application currently deployed online at  https://metrics-as-scores.ml/  is  also functioning as expected with the Qualitas.class data set).

Still, I have several further comments and suggestions—mainly for the JOSS manuscript, but also for the implementation:

1) Given my background in information visualization and visual analytics, I am currently missing any sort of motivation for the design choices regarding the interactive interface (see the section “MAS – The Interactive Application”). The basic question here is: why are line plots used as the visual representations of distributions? The answer for this could be quite straightforward (related to conventions, authors’ and target users’ expectations, prior work, pragmatic considerations, etc.), but in my opinion, such motivation should appear in the manuscript.

2) The manuscript provides a glimpse into the initial motivation behind designing this approach and applying it within the context of software quality metrics research; however, since MAS is positioned as a more general tool, I would strongly suggest extending the “Applications” section within the manuscript with (at least) the list of the currently included/supported data sets (including Iris, etc.), ideally with a very brief discussion of why and how would these data sets be analyzed within MAS.

3) With respect to related work, I would suggest to extend the discussion with the following studies and tools:
- CorpusVis (https://doi.org/10.1109/VISSOFT.2019.00020): visual exploration of software metrics in the Qualitas Corpus
- moreThanANOVA (https://doi.org/10.1371/journal.pone.0271185)
- seaborn (https://doi.org/10.21105/joss.03021): support for numerous statistical visual representations)

4) Several notes related not only to the manuscript, but also the implementation and repository documentation (could be addressed directly or considered as part of future work, for instance):
4.1) While the current Bokeh distribution plot implementation is quite straightforward, I would strongly suggest extending the contents of the tooltip (even visible as part of Figure 1) with the group/domain value (currently must be checked separately within the main plot legend, and would not work that well when a large number of groups and thus colors/hues is present).
4.2) In the “Applications” section, the following is mentioned: “In addition, some of the software metrics in the corpus are never similar across application domains and must be applied with great care when used in quality models (Hönel et al., 2022).”  — while this is mainly relevant to the particular software quality data set, I would suggest mentioning such undesirable choices briefly as part of the data set description within the tool UI (around the “Description” section under “Loaded Dataset”).

5) Minor presentation issues and potential improvements within the manuscript:
5.1) line 17: “allows to assess” — please see https://english.stackexchange.com/a/196130 , https://english.stackexchange.com/q/60271 , and https://ell.stackexchange.com/q/11193
5.2) lines 39+: I would suggest a brief addition at the end of the first section (just one or two sentences) that would provide the readers with an overall idea for how to install and run the tool (i.e., installation with common Python command-line tools + an interactive terminal app + a web-based interactive exploration tool)
5.3) line 75: “which provides access to the PDF/PMF, CDF/CCDF (for scores), and the PPF”  — please expand the abbreviations/acronyms on first use
5.4) lines 79–80: “Cramér–von Mises (Cramér, 1928) and Kolmogorov–Smirnov one-sample (Stephens, 1974) tests, Cramér–von Mises (Anderson, 1962), …” — please check if two different references should indeed be used for Cramér–von Mises tests
5.5) Please check the bibliography carefully for ensuring a consistent style, especially with respect to the capitalization within the publication titles (“Deriving metric thresholds from benchmark data” vs “On the Distribution of the Two-Sample Cramer-von Mises Criterion”; notice “Pymoo: Multi-objective optimization in python” with lowercased “python”, among others…) as well as consistent publication venue titles (“26th IEEE International Conference on Software Maintenance (ICSM 2010)” vs “Proceedings of the 2019 27th ACM Joint Meeting on European Software Engineering Conference and Symposium on the Foundations of Software Engineering”, for instance).
5.5.1) Please double-check the author names: “Carleton, W. A., Anita D.; Florac.”
5.5.2) Please double-check the author names: “John M. Chambers, R. M. H., Anne E. Freeny” (surnames mentioned first across the currently used bibliographical style).

**Comment by kostiantyn-kucher** [comment]
Additionally, please notice the following typo in the repository documentation under "Web Application":  "It feates".

**Comment by MrShoenel** [comment]
1. As you suggested, this design choice is indeed quite straightforward. I have added a sentence in section "MAS -- The Tool- and Analysis Suite".
2. I have made clear that studying the Qualitas.class corpus was the *initial* purpose of MAS. A list of the existing other datasets, including a brief discussion, was added.
3.
  - CorpusVis: Thanks for making me aware of this, I had not seen this before. The goal of that work is different from ours, but still related and relevant. Note that they study the ordinary corpus, not the compiled one though. I have added it to the related works section.
  - moreThanANOVA: This is somewhat related (the systematic approach to comparing quantities is). We have decided to add a reference in the introduction, rather than related work.
  - seaborn: seaborn is a visualization framework that is not actually used by MAS. Instead, we make use of bokeh, which is also cited. seaborn does not offer any other form of visualization that could have been more appropriate than the one implemented by us. Therefore, we have decided not to include or reference it at this point.
4. 
  - 4.1) The tooltip now shows the X/Y coordinates (values), as well as the group's name that is associated with the currently hovered line.
  - 4.2) After discussion, we decided to keep these brief results, as MAS was principally created for the QCC dataset. Also, since we added a brief discussion for the other datasets, we feel that this should stay.
5.
  - 5.1) Changed to "allows the user to assess"
  - 5.2) We have added this and made clear that the application/package is self-sustained.
  - 5.3) Those abbreviations are now properly introduced.
  - 5.4) Both references are indeed required here, as the first one refers to the one-sample test and the second one to the two-sample test (which is an extension made by Anderson).
  - 5.5) We have manually checked every reference now. We used DBLP to obtain precise bibtex records and, if possible/available, even used the publisher's how-to-cite guidelines and the published paper as authoritative guidelines.
    - 5.5.1) That was corrected now!
    - 5.5.2) Also corrected now!

The additional issue of the misspelling in the repository documentation was corrected.

I will leave this issue open for you to close, once you are satisfied with out responses.

**Comment by kostiantyn-kucher** [comment]
Thank you for the updates! I would say that most of my concerns and suggestions have been addressed.

Several further suggestions for minor improvements for the manuscript (I would not insist on these and could close the issue instead, if you prefer; but I believe these should be very easy for you to address):

1. Line 14 (in the PDF generated on July 16): "data visualization and -exploration" — I guess the dash is a typo here...
2. Figure 1: I believe this is still the old screenshot, as I do not see the group title within the tooltip. I would suggest updating the screenshot (with a visible tooltip), say, from the version currently deployed at https://metrics-as-scores.ml/
3. Regarding the bibliography, please note that the respective publishers' guidelines, etc. are not ideal and might contradict each other. I would suggest aiming for consistency in the resulting JOSS manuscript, not the contradictory guidelines and imperfect Bibtex entries... I would strongly suggest to at least edit the following parts of the references (w.r.t. capitalization): "... european ... tallin, estonia, august ..." (Caulo 2019); "... java ... asia pacific ... sydnes, australia, november 30 - december..." (Tempero 2010).

**Comment by MrShoenel** [comment]
Thanks for your feedback, I think you make a valid point about the references, so I am going to fix that (and the other things, too)!

**Comment by MrShoenel** [comment]
I have made some changes to address your points:

1. This was not a typo, it is a hyphenated compound, essentially shortening the two concepts of "data visualization" and "data exploration" into "data visualization and -exploration", in order to avoid repetition of the word "data". Without the hyphen, those would be the two separate concepts of "data visualization" and "exploration".
2. I have updated the screenshot to reflect the latest version of the application. In the meantime, it appears that .ml domains [do no longer work](https://forum.infinityfree.net/t/all-ml-domains-are-down-due-to-registry-transfer/79379), so I have removed references to the hosted web application as well.
3. The venue titles now all follow the same common format. It was a bit tricky to enforce proper capitalization as the generation process for the paper's PDF is different from the one I use locally (where capitalization was correct).
